packages/hebao_v1/contracts/modules/core/UpgraderModule.sol

```diff
@@ -4,32 +4,49 @@ import "../../iface/Controller.sol";
 
 import "../../base/BaseModule.sol";
 
+import "../../thirdparty/OwnedUpgradabilityProxy.sol";
+
 
 /// @title UpgraderModule
-/// @dev This module removes obsolsted modules and add new modules, then
+/// @dev This module removes obsoleted modules and add new modules, then
 ///      removes itself.
 ///
 /// @author Daniel Wang - <daniel@loopring.org>
 ///
 /// The design of this contract is inspired by Argent's contract codebase:
 /// https://github.com/argentlabs/argent-contracts
 contract UpgraderModule is BaseModule {
-    address[] public modulesToRemove;
-    address[] public modulesToAdd;
+    Controller public controller;
+    address[]  public modulesToRemove;
+    address[]  public modulesToAdd;
+    address    public implementation;
 
     constructor(
-        address[] memory _modulesToRemove,
-        address[] memory _modulesToAdd
+        Controller       _controller,
+        address          _implementation,
+        address[] memory _modulesToAdd,
+        address[] memory _modulesToRemove
         )
         public
     {
-        modulesToRemove = _modulesToRemove;
+        controller = _controller;
+        implementation = _implementation;
         modulesToAdd = _modulesToAdd;
+        modulesToRemove = _modulesToRemove;
     }
 
     function activate()
         external
     {
+        if (implementation != OwnedUpgradabilityProxy(msg.sender).implementation() &&
+            controller.implementationRegistry().isImplementationRegistered(implementation)) {
```

Comment:
Because the implementation is set in the module constructor, and only modules that are registered can be used by users, maybe we can just remove the `implementationRegistry` now because it's not really necessary any more?

Reply:
done

Reply:
done

Reply:
done